### PR TITLE
refactor(compiler): apply component metadata asynchronously when defer blocks are present

### PR DESCRIPTION
**Pull request by AndrewKushnir**
This commit updates compiler logic to generate the `setClassMetadataAsync` calls for components that used defer blocks. The `setClassMetadataAsync` function loads deferrable dependencies and invokes the `setClassMetadata` synchronously once everything is loaded. This change is needed to avoid eager references to deferrable symbols in component metadata in generated code.

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by AndrewKushnir** [comment]
Currently blocked by PR #51171.

File: packages/core/src/render3/jit/directive.ts [hunk 1]
```diff
@@ -59,6 +60,12 @@ export function compileComponent(type: Type<any>, metadata: Component): void {
 
   let ngComponentDef: ComponentDef<unknown>|null = null;
 
+  if (getAsyncClassMetadata(type)) {
```
Comment (alxhub) [inline packages/core/src/render3/jit/directive.ts:63]: `compileComponent` is JIT compilation and should have nothing to do with async metadata. Note that this is production code, so it's odd that it refers to TestBed.

Instead, it should be `TestBed` which complains only if you try to override metadata on a component that had unresolved async metadata.

**Review by alxhub — CHANGES_REQUESTED** [review]

File: packages/core/src/render3/jit/directive.ts (hunk 1 above)
Comment (AndrewKushnir) [reply packages/core/src/render3/jit/directive.ts]: Thanks Alex, that makes sense 👍 I've moved this check to the TestBed at the location where we invoked `compileComponent` function.

**Comment by AndrewKushnir** [comment]
@alxhub thanks for the review, the feedback is now addressed, please take a look when you get a chance.

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:553945808:BASE:553945846:1691192451072:225c5304).

**Comment by AndrewKushnir** [comment]
[Global Presubmit](https://test.corp.google.com/OCL:553945808:BASE:554678779:1691466432014:134fa59d).

**Comment by AndrewKushnir** [comment]
[Global Presubmit](https://test.corp.google.com/OCL:553945808:BASE:555342377:1691634234922:7bba6a60).

File: packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts [hunk 2]
```diff
@@ -999,16 +1002,28 @@ export class ComponentDecoratorHandler implements
...
     }
 
     const meta: R3ComponentMetadata<R3TemplateDependency> = {...analysis.meta, ...resolution};
     const fac = compileNgFactoryDefField(toFactoryMetadata(meta, FactoryTarget.Component));
+
+    if (analysis.classMetadata) {
+      // Drop references to existing imports for deferrable symbols that should be present
+      // in the `setClassMetadataAsync` call. Otherwise, an import declaration gets retained.
+      const deferrableSymbols = new Set(deferrableTypes.keys());
+      const rewrittenDecoratorsNode = removeIdentifierReferences(
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:1018]: Aren't we doing this via `setAsyncClassMetadata` anyway? Why would we need to remove the deferred symbols here?

This seems like we're changing the meaning of the user's decorator, which we should try and avoid.

Comment (AndrewKushnir) [reply packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:1018]: We are not removing symbols themselves, we replace existing identifier objects with new ones that do not have connections to the import declarations, [see function logic here](https://github.com/angular/angular/blob/main/packages/compiler-cli/src/ngtsc/annotations/common/src/metadata.ts#L162-L176) (function name is a bit confusing). The decorator shape remains the same. We also perform a very similar operation in [another place](https://github.com/angular/angular/blob/main/packages/compiler-cli/src/ngtsc/annotations/common/src/metadata.ts#L39-L49) (also related to class metadata).

**Review by alxhub — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
**Caretaker note**: TGP is "green", this PR is ready for merge.

**Pull request closed without merging**